app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,48 @@
+class PasswordResetsController < ApplicationController
+  require_unauthenticated_access
+  before_action :ensure_manual_login_allowed
+
+  layout "public"
+
+  def new
+  end
+
+  def create
+    os = request.operating_system
+    browser = request.browser
+    SendResetPasswordEmailJob.perform_later(params[:email], os, browser) # queue as a job to avoid timing attacks
```

Comment:
What's a timing attack, do you mean rate limiting? I just remembered seeing something in the campfire source code for rate limiting:
```
rate_limit to: 10, within: 3.minutes, only: :create, with: -> { render_rejection :too_many_requests }
```

I didn't know this was a feature in rails (and we should check if we're on the correct version). This may be worth adding to the controller.

Reply:
rate limiting would definitely slow down a timing attack, but I'm not sure it completely solves it. Timing attacks are when an attacker figures out if an email (or other) exists for a user in a system by timing how long it takes to get a response. Slower responses indicate that a user is present in the system. If an attacker knows that an email exists then they can use leaked passwords from other websites to gain access to this site since people often use the same password. There's a bunch of other stuff around this and probably much better explanations elsewhere, but that's what I understand anyway.

Probably not a bad idea to do a pass at some rate limiting in the controllers too? Or maybe it's better at the infrastructure layer, but that would mean anyone deploying HostedGPT would have to do it themselves. I wonder if we can just do some generic rate limiting in the ApplicationController and call it good enough

Reply:
Ah, interesting. Now I get what you mean by timing attack. Security isn't a big strength of mine so that's helpful.

On the rate limit, I think we go ahead and add a line like the one below just to this controller and over time we can add them to the others that make sense. I'd hesitate to add it to all of them in one place because my hunch is that we'll need to individualize it for a few endpoints.

Reply:
oh I just looked into this and looks like it's not out in a stable release yet. I think it's in rails 7.2, but that's still a beta release. Unless we want to use the beta release, I'll just leave this out for now. I don't think it's imperative to have just yet